browser/ui/ai_rewriter/ai_rewriter_dialog_delegate.cc

```diff
@@ -86,6 +97,137 @@ class AIRewriterDialogDelegate::DialogContentsObserver
   const raw_ptr<AIRewriterDialogDelegate> dialog_;
 };
 
+class AIRewriterDialogDelegate::DialogPositioner
+    : public views::WidgetObserver,
+      public web_modal::ModalDialogHostObserver {
+ public:
+  DialogPositioner(content::WebContents* target_contents,
+                   web_modal::WebContentsModalDialogHost* host,
+                   views::Widget* dialog_widget)
+      : top_(host->GetDialogPosition({}).y()),
+        target_contents_(target_contents->GetWeakPtr()),
+        dialog_widget_(dialog_widget->GetWeakPtr()) {
+    // TODO(fallaciousreasoning): In a follow up PR we should handle reparenting
+    // |target_contents_| into another browser. For now, we just assume it
+    // remains in the same widget (which it won't necessarily).
+
+    auto* browser = chrome::FindBrowserWithTab(target_contents);
+    CHECK(browser);
+
+    auto* host_widget = views::Widget::GetWidgetForNativeWindow(
+        browser->window()->GetNativeWindow());
+    CHECK(host_widget);
+
+    host_widget_ = host_widget->GetWeakPtr();
+
+    host_widget_observation_.Observe(host_widget);
+    dialog_widget_observation_.Observe(dialog_widget);
+    host_observation_.Observe(host);
+
+    UpdatePosition(target_contents->GetFocusedFrame());
+  }
+
+  DialogPositioner(const DialogPositioner&) = delete;
+  DialogPositioner& operator=(const DialogPositioner&) = delete;
+  ~DialogPositioner() override = default;
+
+  void OnPositionRequiresUpdate() override {
+    if (!target_contents_) {
+      return;
+    }
+
+    UpdatePosition(target_contents_->GetFocusedFrame());
+    UpdateFocusedBounds();
```

Comment:
Q. I guess calling this will end up calling UpdatePosition() again. Is this intended for responsiveness?

Reply:
Yeah it will - if the size of the window changes the position of the element in the bounds almost certainly will too. 

1. The first `UpdatePosition` ensures the dialog stays perfectly in place while the window is being moved around
2. The second one checks if the position of the focused element has moved

Its basically an eager update while we're dragging (which will be correct, if the window hasn't moved).